pubsub/stanpubsub/stan.go

```diff
@@ -0,0 +1,645 @@
+}
+
+// SubscriptionOptions sets options for constructing a *pubsub.Subscription backed by NATS Streaming.
+type SubscriptionOptions struct {
+	// Queue defines the subscription queue name.
+	// See: https://docs.nats.io/developing-with-nats-streaming/queues for more information.
+	Queue string
+	// DurableName,defines the durable subscription name.
+	// See: https://docs.nats.io/developing-with-nats-streaming/durables for more information.
+	DurableName string
+	// MaxInflight controls the maximum number of messages the cluster will have inflight without an ACK.
+	// See: https://docs.nats.io/developing-with-nats-streaming/acks#max-in-flight for more information.
+	MaxInflight int
+	// AckWait defines the duration the cluster will wait for an ACK for a given message.
+	AckWait time.Duration
+	// StartPosition defines the position where the subscription would start..
+	// By default it sets the position to take only new messages.
+	StartAt StartPosition
+	// StartSequence is an option used together with the StartPositionSequenceStart
+	// which defines what is the sequence start position.
+	StartSequence uint64
+	// StartTime is the option used together with the StartPositionTimeStart
+	// which sets the desired start time position and state.
+	StartTime time.Time
+	// ManualAcks is an option that requires the user to manually ack pubsub.Messages.
+	ManualAcks bool
+}
+
+// Validate checks if the subscription options are valid.
+func (s *SubscriptionOptions) Validate() error {
+	if s.StartSequence != 0 && s.StartAt != StartPositionSequenceStart {
+		return errors.New("stanpubsub: subscription option StartSequence should be paired with the StartAt = StartPositionSequenceStart")
+	}
+	if !s.StartTime.IsZero() && s.StartAt != StartPositionTimeStart {
+		return errors.New("stanpubsub: subscription option StartTime should be paired with the StartAt = StartPositionTimeDeltaStart")
+	}
+	return nil
+}
+
+// StartWithLastReceived sets start position to last received.
+func (s *SubscriptionOptions) StartWithLastReceived() {
+	s.StartAt = StartPositionLastReceived
+}
+
+// DeliverAllAvailable sets the start subscription position to the first available message.
+// This should result in delivery of all available messages.
+func (s *SubscriptionOptions) DeliverAllAvailable() {
+	s.StartAt = StartPositionFirst
+}
+
+// StartAtTime sets the desired start time position and state.
+func (s *SubscriptionOptions) StartAtTime(ts time.Time) {
+	s.StartAt = StartPositionTimeStart
+	s.StartTime = ts
+}
+
+// StartAtSequence starts the subscription at given sequence number.
+func (s *SubscriptionOptions) StartAtSequence(sequence uint64) {
+	s.StartAt = StartPositionSequenceStart
+	s.StartSequence = sequence
+}
+
+// StartPosition is the enum that defines start position of subscription.
+type StartPosition int
+
+const (
+	// StartPositionNewOnly is the default subscription start position which delivers only new messages.
+	StartPositionNewOnly StartPosition = iota
```

Comment:
Let's set these to explicit values (0, 1, etc.) because they'll be exposed as part of the API (for URLs). I.e., we wouldn't want the insertion of a new enum value in the middle to renumber everything.